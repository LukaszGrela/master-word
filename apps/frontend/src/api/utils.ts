import { TPartialGameState } from './types';

export const createGameState = (
  attempts: number,
  wordLength: number
): TPartialGameState => {
  const word = Array.from(Array(wordLength)).map(() => '');
  return Array.from(Array(attempts)).map(() => ({
    word: word.concat(),
  }));
};
