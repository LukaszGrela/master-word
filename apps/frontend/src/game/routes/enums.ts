export enum EPaths {
  ROOT = '/',
  GAME = '/game',
  RESULTS = '/results/:result',
}

export const getResultsPath = (result: 'win' | 'lose') => `/results/${result}`;
