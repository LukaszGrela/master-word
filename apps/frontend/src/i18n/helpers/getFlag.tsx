import React from 'react';

export const getFlag = (language: string): React.ReactNode => {
  switch (language) {
    case 'pl':
      return <>ğŸ‡µğŸ‡±</>;
    case 'en':
      return <>ğŸ‡ºğŸ‡¸</>;
    default:
      return <>{language.toUpperCase()}</>;
  }
};
