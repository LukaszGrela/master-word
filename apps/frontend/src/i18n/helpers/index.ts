/* v8 ignore start */
export * from './loadTranslation';
/* v8 ignore stop */
