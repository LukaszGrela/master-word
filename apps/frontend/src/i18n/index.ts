/* v8 ignore start */
export * from './LanguageContext/hooks';
export { LanguageProvider } from './LanguageContext/index';
export type { ILanguageContext } from './LanguageContext/index';
/* v8 ignore stop */
