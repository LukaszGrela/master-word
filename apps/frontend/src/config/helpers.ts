import { AppStorage, EStorageKeys } from '@repo/utils';
import { IConfig } from './types';
import { ATTEMPTS, LANGUAGE, WORD_LENGTH } from '../game';

/**
 * Assure local storage is using enabled values from current config object
 * @param config Current Configuration
 */
export function updateLocalStorageConfig(config: IConfig) {
  console.log('config', config);
  const storage = AppStorage.getInstance();
  // game language
  const configLang =
    config.enabledLanguages.length > 0 ? config.enabledLanguages : [LANGUAGE];
  if (!storage.has(EStorageKeys.GAME_LANGUAGE)) {
    storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
  } else {
    // value exists, confirm it is allowed
    if (
      configLang.indexOf(storage.getItem(EStorageKeys.GAME_LANGUAGE)!) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
    }
  }

  // game word length
  const configLength =
    config.enabledLength.length > 0 ? config.enabledLength : [WORD_LENGTH];
  if (!storage.has(EStorageKeys.GAME_WORD_LENGTH)) {
    storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configLength.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_WORD_LENGTH)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
    }
  }

  // game attempts
  const configAttempts =
    config.enabledAttempts.length > 0 ? config.enabledAttempts : [ATTEMPTS];
  if (!storage.has(EStorageKeys.GAME_ATTEMPTS)) {
    storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configAttempts.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_ATTEMPTS)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
    }
  }

  // UI
  if (!storage.has(EStorageKeys.UI_LANGUAGE)) {
    storage.setItem(EStorageKeys.UI_LANGUAGE, LANGUAGE);
  }
}
