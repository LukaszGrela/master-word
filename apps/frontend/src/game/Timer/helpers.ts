import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
