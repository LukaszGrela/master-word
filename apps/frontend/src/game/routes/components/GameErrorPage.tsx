import {
  Location,
  useLocation,
  useNavigate,
  useRouteError,
} from 'react-router-dom';
import { TGameSession } from '@repo/backend-types';
import { useLanguage } from '../../../i18n';
import { GameLanguage } from '../../language';
import { useCallback } from 'react';
import { EPaths } from '../enums';
import { TErrorResponse } from '../../../api';

export const GameErrorPage = () => {
  const { getUIText: t } = useLanguage();
  const navigate = useNavigate();
  const location = useLocation() as Location<
    { session: TGameSession; error: TErrorResponse } | undefined
  >;
  const gameSession = location.state?.session;
  const gameError = location.state?.error;

  const routerError = useRouteError() as {
    statusText?: string;
    message?: string;
  };

  console.error(location.state?.error);
  console.error(routerError);
  const handleAction = useCallback(
    (action: 'again' | 'quit') => () => {
      if (action === 'quit') {
        navigate(EPaths.ROOT, { replace: true });
      }
      if (action === 'again') {
        navigate(EPaths.GAME, {
          replace: true,
          state: gameSession?.session,
        });
      }
    },
    [gameSession?.session, navigate],
  );

  return (
    <div className="game-error">
      <h1>{t('general-error-title')}</h1>
      <h2>{t('game-error-subtitle')}</h2>
      <p>{t('general-error-message')}</p>
      {gameError && (
        <p>
          <i>Code: {gameError.code}</i> <i>Error: {gameError.error}</i>
        </p>
      )}
      {routerError && (
        <p>
          <i>{routerError.statusText || routerError.message}</i>
        </p>
      )}
      {gameSession && <GameLanguage />}
      <div className="button-row">
        <button onClick={handleAction('quit')} className={'secondary'}>
          {t('general-error-btn-home')}
        </button>
        {gameSession && (
          <button onClick={handleAction('again')} className={'primary'}>
            {t('result-again-button')}
          </button>
        )}
      </div>
    </div>
  );
};

/* v8 ignore start */
export function Component() {
  return <GameErrorPage />;
}
Component.displayName = 'LazyGameErrorPage';
/* v8 ignore stop */
