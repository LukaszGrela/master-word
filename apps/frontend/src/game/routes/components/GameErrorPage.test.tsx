import { Mock, afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { defaultLanguageContext } from '../../../i18n/LanguageContext/__tests__/contextRenderer';
import { GameErrorPage } from './GameErrorPage';
import { AppStorage, IStorage, hasOwn } from '@repo/utils';
import { IDictionary } from '@repo/common-types';
import { replaceSubstituteMap } from '../../../i18n/helpers';
import { createMemoryRouter } from 'react-router-dom';
import { EPaths } from '../enums';
import { renderMockRouter } from './__tests__/helpers';
import { fireEvent, waitFor } from '@testing-library/react';

vi.mock('@repo/utils', async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
  const original = await importOriginal<any>();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const AppStorageMock = vi.fn<any, IStorage>();
  const proto = AppStorageMock.prototype as IStorage;

  proto.getItem = vi.fn();
  proto.setItem = vi.fn();
  proto.removeItem = vi.fn();
  proto.has = vi.fn();
  proto.clear = vi.fn();
  proto.clearStorage = vi.fn();

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const instance = new AppStorageMock();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
  (AppStorageMock as any).getInstance = vi.fn().mockReturnValue(instance);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...original,

    AppStorage: AppStorageMock,
  };
});

const originalFetch = global.fetch;

describe('game', () => {
  describe('routes', () => {
    describe('components', () => {
      describe('GameErrorPage', () => {
        const translationPl: IDictionary<string> = {
          'general-error-title': 'Ożesz kurczaczki!',
          'general-error-message':
            'Przepraszamy, nieoczekiwany błąd jak piorun z nieba spadł i zepsuł zabawę...',
          'general-error-btn-home': 'Zamknij',
          'game-error-subtitle': 'Problem z sesją gry.',
          'result-again-button': 'Odgadnij następne słowo',
          'translation-info-sr': 'Wybierz język interfejsu',
          'translation-button-en': 'Angielski',
          'translation-button-pl': 'Polski',
          'result-language-selector-btn-title-pl': 'Słowo polskie',
          'result-language-selector-btn-title-en': 'Słowo angielskie',
        };
        let mockFetch = vi.fn();
        let storage: IStorage = AppStorage.getInstance();
        beforeEach(() => {
          mockFetch = vi.fn();
          storage = AppStorage.getInstance();

          (
            storage.getItem as Mock<[key: string], string | null>
          ).mockImplementation((key) => {
            if (key === 'ui-language') {
              return 'pl';
            }
            if (key === 'word-language') {
              return 'pl';
            }
            return null;
          });

          global.fetch = mockFetch;
        });
        afterEach(() => {
          vi.clearAllMocks();
          global.fetch = originalFetch;
        });
        it('renders', async () => {
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <p>HOME</p>,
            },
            {
              path: EPaths.GAME_ERROR,
              element: <GameErrorPage />,
            },
          ]);
          await router.navigate(EPaths.GAME_ERROR, {
            state: {
              session: { session: 'session' },
              error: { code: 404, error: 'Not found' },
            },
          });

          const { container } = renderMockRouter(router, {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          expect(container).toMatchSnapshot();
        });
        it('renders - play again', async () => {
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <p>HOME</p>,
            },
            {
              path: EPaths.GAME_ERROR,
              element: <GameErrorPage />,
            },
          ]);
          await router.navigate(EPaths.GAME_ERROR, {
            state: {
              session: { session: 'session' },
              error: { code: 404, error: 'Not found' },
            },
          });

          const { getByText } = renderMockRouter(router, {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          fireEvent.click(getByText(translationPl['result-again-button']));

          await waitFor(() => {
            expect(router.state.location.pathname).toEqual('/game');
            expect(router.state.location.state).toBe('session');
          });
        });
        it('renders without game session', async () => {
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <p>HOME</p>,
            },
            {
              path: EPaths.GAME_ERROR,
              element: <GameErrorPage />,
            },
          ]);
          await router.navigate(EPaths.GAME_ERROR, {
            state: {
              error: { code: 404, error: 'Not found' },
            },
          });

          const { container } = renderMockRouter(router, {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          expect(container).toMatchSnapshot();
        });
        it('renders without game session - close', async () => {
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <p>HOME</p>,
            },
            {
              path: EPaths.GAME_ERROR,
              element: <GameErrorPage />,
            },
          ]);
          await router.navigate(EPaths.GAME_ERROR, {
            state: {
              error: { code: 404, error: 'Not found' },
            },
          });

          const { getByText } = renderMockRouter(router, {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          fireEvent.click(getByText(translationPl['general-error-btn-home']));

          await waitFor(() => {
            expect(router.state.location.pathname).toEqual('/');
            expect(router.state.location.state).toBeNull();
          });
        });
      });
    });
  });
});
