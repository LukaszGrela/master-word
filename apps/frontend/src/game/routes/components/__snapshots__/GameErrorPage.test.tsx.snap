// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > routes > components > GameErrorPage > renders 1`] = `
<div>
  <div
    class="game-error"
  >
    <h1>
      Ożesz kurczaczki!
    </h1>
    <h2>
      Problem z sesją gry.
    </h2>
    <p>
      Przepraszamy, nieoczekiwany błąd jak piorun z nieba spadł i zepsuł zabawę...
    </p>
    <p>
      <i>
        Code: 
        404
      </i>
       
      <i>
        Error: 
        Not found
      </i>
    </p>
    <div
      class="game-language"
    >
      <p
        class="language-info"
      >
        result-language-info
      </p>
      <div
        class="language-selector translation"
      >
        <span
          class="hidden sr"
        >
          Wybierz język interfejsu
        </span>
        <button
          class="translation-btn selected"
          title="Słowo polskie"
        >
          <span
            class="label"
          >
            word-language-selector-btn-label-pl
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/pl.png"
              srcset="https://flagcdn.com/w40/pl.png 2x"
              width="20"
            />
          </span>
        </button>
        <button
          class="translation-btn"
          title="Słowo angielskie"
        >
          <span
            class="label"
          >
            word-language-selector-btn-label-en
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/us.png"
              srcset="https://flagcdn.com/w40/us.png 2x"
              width="20"
            />
          </span>
        </button>
      </div>
    </div>
    <div
      class="button-row"
    >
      <button
        class="secondary"
      >
        Zamknij
      </button>
      <button
        class="primary"
      >
        Odgadnij następne słowo
      </button>
    </div>
  </div>
</div>
`;

exports[`game > routes > components > GameErrorPage > renders without game session 1`] = `
<div>
  <div
    class="game-error"
  >
    <h1>
      Ożesz kurczaczki!
    </h1>
    <h2>
      Problem z sesją gry.
    </h2>
    <p>
      Przepraszamy, nieoczekiwany błąd jak piorun z nieba spadł i zepsuł zabawę...
    </p>
    <p>
      <i>
        Code: 
        404
      </i>
       
      <i>
        Error: 
        Not found
      </i>
    </p>
    <div
      class="button-row"
    >
      <button
        class="secondary"
      >
        Zamknij
      </button>
    </div>
  </div>
</div>
`;
